onnxruntime/core/optimizer/nhwc_transformer.cc

```diff
@@ -0,0 +1,366 @@
+  std::string new_output_def_name = graph_.GenerateNodeArgName("output");
+  auto* new_output_arg = &graph_.GetOrCreateNodeArg(new_output_def_name, output_defs[0]->TypeAsProto());
+
+  Node& permute_node = graph_.AddNode(graph_.GenerateNodeName(bNHWC ? "PermuteNCHW" : "PermuteNHWC"),
+                                     bNHWC ? "ReorderIntput" : "ReorderOutput",
+                                     bNHWC ? "ReorderIntput" : "ReorderOutput",
+                                     {new_output_arg},
+                                     {output_defs[0]},
+                                     nullptr,
+                                     kMSNhwcDomain);
+  permute_node.SetExecutionProviderType(provider_);
+  output_defs[0] = new_output_arg;
+  graph_.AddEdge(node.Index(), permute_node.Index(), 0, 0);
+
+  return permute_node.Index();
+}
+
+bool NhwcTransformerImpl::RequiresWeightsPermutation(const Node& node) {
+   return ((node.GetExecutionProviderType() == kAclExecutionProvider ||
+            node.GetExecutionProviderType() == kArmNNExecutionProvider) &&
+           (node.OpType() == "Conv" ||
+            node.OpType() == "FusedConv"));
+}
+
+NodeIndex NhwcTransformerImpl::ReplaceNode(Node& node) {
+
+  auto& input_defs = node.MutableInputDefs();
+  auto& output_defs = node.MutableOutputDefs();
+
+  LOGS(logger, VERBOSE) << "NHWC >> Replace " << node.OpType();
+
+  Node& newNode = graph_.AddNode(graph_.GenerateNodeName(node.Name()),
+                                     node.OpType(),
+                                     node.OpType(),
+                                     input_defs,
+                                     output_defs,
+                                     &node.GetAttributes(),
+                                     kMSNhwcDomain);
+  newNode.SetExecutionProviderType(node.GetExecutionProviderType());
+
+  // Permute weights
+  if (RequiresWeightsPermutation(node))
+    PermuteWeights(input_defs[1], &newNode.MutableInputDefs()[1], node.GetExecutionProviderType());
+
+  NodeIndex oldIndex = node.Index();
+  const std::vector<std::reference_wrapper<Node>> replacedNode({node});
+  graph_utils::FinalizeNodeFusion(graph_, replacedNode, newNode);
+  nodes_layout.erase(oldIndex);
+
+  return newNode.Index();
+}
+
+void NhwcTransformerImpl::PermuteWeights(NodeArg *input_def, NodeArg** nhwc_conv_W_arg, __attribute__ ((unused)) const std::string& execution_provider) {
+  // Require that the weights tensor be static.
+  const ONNX_NAMESPACE::TensorProto* conv_W_tensor_proto = nullptr;
+  if (!graph_.GetInitializedTensor(input_def->Name(), conv_W_tensor_proto) ||
+      (conv_W_tensor_proto->data_type() != ONNX_NAMESPACE::TensorProto_DataType_FLOAT) ||
+      (conv_W_tensor_proto->dims_size() != 4) ||
+      (execution_provider == kArmNNExecutionProvider && conv_W_tensor_proto->dims(1) == 1)) {
+
+    return;
+  }
+
+  Initializer conv_W{*conv_W_tensor_proto, graph_.ModelPath()};
+
+  arm_compute::Tensor weights;
```

Comment:
To my comment about possibly doing this in MLAS too, removing the dependency on ACL here would be useful. Is a CPU implementation possible here? Maybe just use Eigen to reorder?

Reply:
I think it can be done as long as we know the layout required by the library/accelerator behind the EP. The performance for the graph transformation step is not critical.